### PR TITLE
Allow accounting for direct-to-zone DOAS on zone sizing.

### PR DESCRIPTION
**Pull request by wfbuhl**
An addition to Sizing:Zone input allows the user to specify that a direct-to-zone Dedicated Outdoor Air System's heating or cooling effect needs to be added to the zone design heating or cooling loads. The user can specify DOAS supply temperatures and some limited information about the DOAS operation strategy.

### CODE REVIEW

**Comment by mbadams5** [comment]
@wfbuhl There are still more changes to SQLite code needed for the new output field. That is why SQLiteFixture.addSQLiteZoneSizingRecord unit test is failing. I can expand on this comment later, or I may just go ahead and fix the necessary issues.

**Comment by wfbuhl** [comment]
Hum - I can take a look at it tomorrow if you don't want to do it, though a
suggestion would always help. At some point I think I forgot about SQL and
only focused on my stuff.

Fred

**Comment by wfbuhl** [comment]
Because of changes to the IDD, the hourly report variables, and the eio the regression tests will fail for all example files. Additionally the predefined reports have changed so the html files will not be the same.

What should be the same are the csv outputs. I have checked some of these on my own machine and they are OK. I can't tell if they are OK in these regression tests but I think that is what they are saying. 

The predefined reports change are as follows:

HVAC Sizing Summary - Zone Cooling and Zone Heating. We've added 2 columns - "minimum outdoor airflow rate" and "heat gain from DOAS". Also "Thermostat setpoint at peak load' was coming out zero for duat setpoint cases and this has now been corrected.

Sensible Heat Gain Summary has been changed: the 1st 2 columns column  have been split into "HVAC Zone Eq & Other Sensible Air Heating" and "HVAC Terminal Unit Sensible Air Heating" and ditto for Cooling.

Lastly Zone Component Load Summary has had a row added: "Doas Direct to Zone".

Fred

**Comment by Myoldmopar** [comment]
I think this fixes #3770.  @wfbuhl Can you confirm?

**Comment by wfbuhl** [comment]
It certainly fixes this:

"If a DOAS is allowed to operate at cold or hot temperatures, and it is
sequenced #1 https://github.com/NREL/EnergyPlus/issues/1, then it can
easily happen that the equipment #2
https://github.com/NREL/EnergyPlus/issues/2 is not sized large enough to
handle the loads from the DOAS. "

It doesn't deal with all possible cases in which eq#1 could affect the size
of eq#2. But it should deal with most DOAS cases.

Fred

**Comment by Myoldmopar** [comment]
@wfbuhl Whenever you get around to fixing the failing build above, make sure to pull in the latest develop so there aren't unrelated diffs from CI.

**Comment by wfbuhl** [comment]
I'm back and just doing it now.

**Comment by wfbuhl** [comment]
So all build warnings are eliminated. Integration and unit tests are OK. The fails are expected but I will double check them. Looks like this is just about ready to go.

**Comment by wfbuhl** [comment]
This is ready to merge

**Comment by Myoldmopar** [comment]
@wfbuhl CI went kinda insane and ~~left you lots of comments~~.  I'm getting the CI machines restarted and cleaned up.  In light of this, I pulled develop into this branch as well and we should get some once-and-for-all good CI results.  If all is well, I'll review and merge.

Update: I cleaned up lots of the unnecessary comments.

**Comment by Myoldmopar** [comment]
@wfbuhl @mjwitte OK, I looked over the source and made some changes that shouldn't affect functionality at all, just cleanup.  I looked at the CI results and confirmed that the changes are only as expected.  Once CI verifies I didn't break anything with my minor commit, I'm OK with this going in.

File: idd/Energy+.idd.in [hunk 1]
```diff
@@ -30616,9 +30616,33 @@ Sizing:Zone,
       \minimum 0
       \default 0.3
       \note This input is used for autosizing components when Heating Design Air Flow Method = DesignDayWithLimit.
-  A7; \field Design Specification Zone Air Distribution Object Name
+  A7, \field Design Specification Zone Air Distribution Object Name
       \type object-list
       \object-list DesignSpecificationZoneAirDistributionNames
+  A8, \field Account for Dedicated Outside Air System
```
Comment (mjwitte) [inline idd/Energy+.idd.in:30622]: @wfbuhl Outside should be Outdoor everywhere for consistency with other field names.

File: idd/Energy+.idd.in [hunk 2]
```diff
@@ -30616,9 +30616,33 @@ Sizing:Zone,
...
+      \default No
+  A9, \field Dedicated Outside Air System Control Strategy
+      \note 1)supply neutral ventilation air; 2)supply neutral dehumidified and reheated 
+      \note ventilation air; 3)supply cold ventilation air
+      \type choice
+      \key NeutralSupplyAir
+      \key NeutralDehumidifiedSupplyAir
+      \key ColdSupplyAir
+      \default NeutralSupplyAir
+  N17,\field Dedicated Outside Air Low Setpoint for Design
```
Comment (mjwitte) [inline idd/Energy+.idd.in:30636]: @wfbuhl This field and the next should say . . . Temperature Setpoint . . .

**Comment by Myoldmopar** [comment]
@wfbuhl Could you provide a "synopsis for users" as the title of this PR, and also maybe edit the first comment to describe the changes in a couple sentences or bullet points?  That way if the user clicks on it in the Changelog and they make it to this page, they'll be able to quickly see the changes brought in by this branch.

Thanks!

**Comment by Myoldmopar** [comment]
@wfbuhl @mjwitte I fixed a line ending issue in the docs and pushed that up.  I was about to say this is ready to go in, but I noticed there are diffs in _SmallOffice_Dulles.  Are they expected?  From a quick glance they might be small, but not completely insignificant.

**Comment by wfbuhl** [comment]
Let me take a look.

Fred

**Comment by wfbuhl** [comment]
I believe it is OK. I changed where AirTerminal:SingleDuct:Uncontrolled().HeatRate etc. is calculated to make it consistent with the other terminal units and with SumMCpDTsystem. This is to get the Sensible Heat Gain Components report to work correctly (the rows must sum to zero) and get my new column to fit in with how the other columns are calculated. Now this stuff is calculated in ZoneTempPredictorCorrector:

for ( NodeNum = 1; NodeNum <= ZoneEquipConfig( ZoneEquipConfigNum ).NumInletNodes; ++NodeNum ) {
				// Get node conditions
				NodeTemp = Node( ZoneEquipConfig( ZoneEquipConfigNum ).InletNode( NodeNum ) ).Temp;
				MassFlowRate = Node( ZoneEquipConfig( ZoneEquipConfigNum ).InletNode( NodeNum ) ).MassFlowRate;
				CpAir = PsyCpAirFnWTdb( ZoneAirHumRat( ZoneNum ), NodeTemp );

				SumMCpDTsystem += MassFlowRate * CpAir * ( NodeTemp - MAT( ZoneNum ) );

			} // NodeNum

			if ( ZoneEquipConfig( ZoneEquipConfigNum ).ADUNum > 0 ) {
				ADUNum = ZoneEquipConfig( ZoneEquipConfigNum ).ADUNum;
				NodeTemp = Node( AirDistUnit( ADUNum ).OutletNodeNum ).Temp;
				MassFlowRate = Node( AirDistUnit( ADUNum ).OutletNodeNum ).MassFlowRate;
				CpAir = PsyCpAirFnWTdb( ZoneAirHumRat( ZoneNum ), NodeTemp );
				ADUHeatAddRate = MassFlowRate * CpAir * ( NodeTemp - MAT( ZoneNum ) );
				AirDistUnit( ADUNum ).HeatRate = max( 0.0, ADUHeatAddRate );
				AirDistUnit( ADUNum ).CoolRate = std::abs( min( 0.0, ADUHeatAddRate ) );
				AirDistUnit( ADUNum ).HeatGain = AirDistUnit( ADUNum ).HeatRate * TimeStepSys * SecInHour;
				AirDistUnit( ADUNum ).CoolGain = AirDistUnit( ADUNum ).CoolRate * TimeStepSys * SecInHour;
			}

			if ( ZoneEquipConfig( ZoneEquipConfigNum ).SDUNum > 0 ) {
				SDUNum = ZoneEquipConfig( ZoneEquipConfigNum ).SDUNum;
				NodeTemp = Node( DirectAir( SDUNum ).ZoneSupplyAirNode ).Temp;
				MassFlowRate = Node( DirectAir( SDUNum ).ZoneSupplyAirNode ).MassFlowRate;
				CpAir = PsyCpAirFnWTdb( ZoneAirHumRat( ZoneNum ), NodeTemp );
				SDUHeatAddRate = MassFlowRate * CpAir * ( NodeTemp - MAT( ZoneNum ) );
				DirectAir( SDUNum ).HeatRate = max( SDUHeatAddRate, 0.0 );
				DirectAir( SDUNum ).CoolRate = std::abs( min( SDUHeatAddRate, 0.0 ) );
				DirectAir( SDUNum ).HeatEnergy = DirectAir( SDUNum ).HeatRate * TimeStepSys * SecInHour;
				DirectAir( SDUNum ).CoolEnergy = DirectAir( SDUNum ).CoolRate * TimeStepSys * SecInHour;
			}




:

**Comment by wfbuhl** [comment]
I believe it is OK. I moved where the output variables of AirTerminal:SingleDuct:Uncontrolled are calculated from DirectAirManager to ZoneTempPredictorCorrector so that they are calculated identically with SumMCpDTsystem. Also moved AirDistUnit().HeatRate etc. to ZoneTempPredictorCorrector. All this to make the Sensible Heat Gain Summary report come out correctly: the columns have to sum to zero and I've split an existing column into two columns. Here is where this stuff is calculated now in ZoneTempPredictorCorrector starting at line 4936:

```
        for ( NodeNum = 1; NodeNum <= ZoneEquipConfig( ZoneEquipConfigNum ).NumInletNodes; ++NodeNum ) {
            // Get node conditions
            NodeTemp = Node( ZoneEquipConfig( ZoneEquipConfigNum ).InletNode( NodeNum ) ).Temp;
            MassFlowRate = Node( ZoneEquipConfig( ZoneEquipConfigNum ).InletNode( NodeNum ) ).MassFlowRate;
            CpAir = PsyCpAirFnWTdb( ZoneAirHumRat( ZoneNum ), NodeTemp );

            SumMCpDTsystem += MassFlowRate * CpAir * ( NodeTemp - MAT( ZoneNum ) );

        } // NodeNum

        if ( ZoneEquipConfig( ZoneEquipConfigNum ).ADUNum > 0 ) {
            ADUNum = ZoneEquipConfig( ZoneEquipConfigNum ).ADUNum;
            NodeTemp = Node( AirDistUnit( ADUNum ).OutletNodeNum ).Temp;
            MassFlowRate = Node( AirDistUnit( ADUNum ).OutletNodeNum ).MassFlowRate;
            CpAir = PsyCpAirFnWTdb( ZoneAirHumRat( ZoneNum ), NodeTemp );
            ADUHeatAddRate = MassFlowRate * CpAir * ( NodeTemp - MAT( ZoneNum ) );
            AirDistUnit( ADUNum ).HeatRate = max( 0.0, ADUHeatAddRate );
            AirDistUnit( ADUNum ).CoolRate = std::abs( min( 0.0, ADUHeatAddRate ) );
            AirDistUnit( ADUNum ).HeatGain = AirDistUnit( ADUNum ).HeatRate * TimeStepSys * SecInHour;
            AirDistUnit( ADUNum ).CoolGain = AirDistUnit( ADUNum ).CoolRate * TimeStepSys * SecInHour;
        }

        if ( ZoneEquipConfig( ZoneEquipConfigNum ).SDUNum > 0 ) {
            SDUNum = ZoneEquipConfig( ZoneEquipConfigNum ).SDUNum;
            NodeTemp = Node( DirectAir( SDUNum ).ZoneSupplyAirNode ).Temp;
            MassFlowRate = Node( DirectAir( SDUNum ).ZoneSupplyAirNode ).MassFlowRate;
            CpAir = PsyCpAirFnWTdb( ZoneAirHumRat( ZoneNum ), NodeTemp );
            SDUHeatAddRate = MassFlowRate * CpAir * ( NodeTemp - MAT( ZoneNum ) );
            DirectAir( SDUNum ).HeatRate = max( SDUHeatAddRate, 0.0 );
            DirectAir( SDUNum ).CoolRate = std::abs( min( SDUHeatAddRate, 0.0 ) );
            DirectAir( SDUNum ).HeatEnergy = DirectAir( SDUNum ).HeatRate * TimeStepSys * SecInHour;
            DirectAir( SDUNum ).CoolEnergy = DirectAir( SDUNum ).CoolRate * TimeStepSys * SecInHour;
        }
```

You can also see the code change on GitHub. The reason for the difference is Cp is different and the zone temperature is slightly different.

**Comment by Myoldmopar** [comment]
Thanks for looking into it @wfbuhl, your justification sounds satisfactory and thus your branch wins and will go in first.  This evening I will try to help folks get all the other PR branches updated with this latest develop so that their branches don't show diffs.

Merging in exactly five minutes :clock5: :clock4: :clock3: :clock2: :clock1: :clock12: (_those aren't minutes..._), unless someone gives me a reason not to, @mjwitte.

**Pull request merged**